plugins/thunderbolt/fu-plugin-thunderbolt.c

```diff
@@ -0,0 +1,736 @@
+}
+
+static guint64
+udev_device_get_sysattr_guint64 (GUdevDevice *device,
+				 const gchar *name,
+				 GError **error)
+{
+	const gchar *sysfs;
+	guint64 val;
+
+	sysfs = g_udev_device_get_sysfs_attr (device, name);
+	if (sysfs == NULL) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_INTERNAL,
+			     "failed get id %s for %s", name, sysfs);
+		return 0x0;
+	}
+
+	val = g_ascii_strtoull (sysfs, NULL, 16);
+	if (val == 0x0) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_INTERNAL,
+			     "failed to parse %s", sysfs);
+		return 0x0;
+	}
+
+	return val;
+}
+
+static guint16
+fu_plugin_thunderbolt_udev_get_id (GUdevDevice *device,
+				   const gchar *name,
+				   GError **error)
+{
+
+	guint64 id;
+
+	id = udev_device_get_sysattr_guint64 (device, name, error);
+	if (id == 0x0)
+		return id;
+
+	if (id > G_MAXUINT16) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_INTERNAL,
+			     "vendor id overflows");
+		return 0x0;
+	}
+
+	return (guint16) id;
+}
+
+static gboolean
+fu_plugin_thunderbolt_is_host (GUdevDevice *device)
+{
+	g_autoptr(GUdevDevice) parent = NULL;
+	const gchar *name;
+
+	/* the (probably safe) assumption this code makes is
+	 * that the thunderbolt device which is a direct child
+	 * of the domain is the host controller device itself */
+	parent = g_udev_device_get_parent (device);
+	name = g_udev_device_get_name (parent);
+	if (name == NULL)
+		return FALSE;
+
+	return g_str_has_prefix (name, "domain");
+}
+
+static void
+fu_plugin_thunderbolt_add (FuPlugin *plugin, GUdevDevice *device)
+{
+	FuDevice *dev_tmp;
+	const gchar *name;
+	const gchar *uuid;
+	const gchar *vendor;
+	const gchar *version;
+	const gchar *devpath;
+	gboolean is_host;
+	guint16 did;
+	guint16 vid;
+	g_autofree gchar *id = NULL;
+	g_autofree gchar *vendor_id = NULL;
+	g_autofree gchar *device_id = NULL;
+	g_autoptr(FuDevice) dev = NULL;
+	g_autoptr(GError) error = NULL;
+
+	uuid = g_udev_device_get_sysfs_attr (device, "unique_id");
+	if (uuid == NULL) {
+		/* most likely the domain itself, ignore */
+		/* TODO: handle devices in safe-mode */
+		return;
+	}
+
+	devpath = g_udev_device_get_sysfs_path (device);
+
+	g_debug ("adding udev device: %s at %s", uuid, devpath);
+
+	id = fu_plugin_thunderbolt_gen_id (device);
+	dev_tmp = fu_plugin_cache_lookup (plugin, id);
+	if (dev_tmp != NULL) {
+		g_debug ("ignoring duplicate %s", id);
+		return;
+	}
+
+	vid = fu_plugin_thunderbolt_udev_get_id (device, "vendor", &error);
+	if (vid == 0x0) {
+		g_warning ("failed to get Vendor ID: %s", error->message);
+		return;
+	}
+	did = fu_plugin_thunderbolt_udev_get_id (device, "device", &error);
+	if (did == 0x0) {
+		g_warning ("failed to get Device ID: %s", error->message);
+		return;
+	}
+
+	is_host = fu_plugin_thunderbolt_is_host (device);
+	vendor_id = g_strdup_printf ("TBT:0x%04X", (guint) vid);
+	device_id = g_strdup_printf ("TBT-%04x%04x", (guint) vid, (guint) did);
```

Comment:
@superm1 I'm thinking more on users with existing platform out there that till now couldn't update the FR FW for having P2P working withiut using Windows and now may like to do it. But yes, the chances that vendors will upload FR FW to LVFS is not very high, to say the least.

For updating AR device that is connected to FR host, I recall that Mika, who developed the linux kernel module, told my that he did it and it worked, but I'm maybe wrong. (BTW, on Windows we don't support it :) )